includes/AMP/Traits/Sanitization_Utils.php

```diff
@@ -114,6 +114,139 @@ private function transform_a_tags( &$document ): void {
 		}
 	}
 
+	/**
+	 * Transforms all paragraphs in a story to use semantic heading tags if needed.
+	 *
+	 * This logic here mirrors the getTextElementTagNames() function in the editor
+	 * in order to change simple <p> tags into <h1>, <h2> or <h3>, depending on font size.
+	 *
+	 * Only relevant for older stories that haven't been updated in a while.
+	 *
+	 * @since 1.18.0
+	 *
+	 * @param Document|AMP_Document $document   Document instance.
+	 * @param bool                  $is_enabled Whether the feature is enabled.
+	 * @return void
+	 */
+	private function use_semantic_heading_tags( &$document, $is_enabled ): void {
+		if ( ! $is_enabled ) {
+			return;
+		}
+
+		$pages = $document->getElementsByTagName( 'amp-story-page' );
+
+		/**
+		 * The <amp-story-page> element
+		 *
+		 * @var DOMElement $page The <amp-story-page> element
+		 */
+		foreach ( $pages as $page ) {
+			$h1 = $page->getElementsByTagName( 'h1' );
+			$h2 = $page->getElementsByTagName( 'h2' );
+			$h3 = $page->getElementsByTagName( 'h3' );
```

Comment:
Instead of doing this, a string search might be less expensive, resource wise.

Reply:
We're operating on a DOMDocument instance here, can't really perform a string search on that.